### ISSUE TITLE
🐞 PaDiM 7x faster and PatchCore 1.5x at inference, identical output: the anomaly map blur isn't applied separably

### ISSUE THREAD

**Issue by andreifagds** [open-edge-platform/anomalib#3780]
### Describe the bug

I was profiling PatchCore and found that the anomaly map generator takes about 40% of the forward pass. Almost all of that is the Gaussian blur, and the blur is doing roughly 16 times more arithmetic than it needs to.

One forward at 320×320, `wide_resnet50_2`, memory bank 8000×1536, batch 1. Measured forward is 60.99 ms; the stages below were timed individually, which adds about 17% of overhead, so they sum to more than the total:

| stage | time | % of forward |
|---|---|---|
| nearest-neighbour search | 35.90 ms | 58.9% |
| **anomaly_map_generator** | **25.38 ms** | **41.6%** |
| feature extractor + pooling | 8.30 ms | 13.6% |
| compute_anomaly_score | 1.71 ms | 2.8% |
| generate_embedding | 0.33 ms | 0.5% |

The upsample inside the map generator costs 0.01 ms, so essentially all of those 25.38 ms are the blur.

Applying the same kernel separably brings the map generator to 0.54 ms. The output is the same. End to end, same config:

| model | today | fixed | speedup |
|---|---|---|---|
| PatchCore | 60.99 ms | 41.70 ms | **1.46x** |
| PaDiM | 30.05 ms | 4.26 ms | **7.06x** |

PaDiM gains far more because it has no nearest-neighbour search to dilute the change, so the blur was most of its forward. PatchCore's own figure moves with memory bank size: with a smaller bank the search gets cheap and I measured up to 3.2x, but 1.46x with a realistic bank is the number I'd trust.

### Why it's slow

Two things pile up here.

The first is that a Gaussian kernel is separable. Blurring along x and then along y gives you the same result as one 2D pass, but it costs `2k` multiply-adds per pixel instead of `k*k`. At the default `sigma=4` that's 66 operations instead of 1089. `GaussianBlur2d` builds the full 2D kernel and runs a single `F.conv2d`.

The second one I only found by accident, sweeping kernel sizes on a plain depthwise conv2d at 320×320:

| kernel | TF32 on | TF32 off | ratio |
|---|---|---|---|
| 17×17 | 0.639 ms | 0.651 ms | 1.0x |
| 25×25 | 15.486 ms | 1.344 ms | 11.5x |
| **33×33** | **26.103 ms** | 2.443 ms | **10.7x** |
| 41×41 | 40.483 ms | 3.429 ms | 11.8x |
| 49×49 | 5.513 ms | 5.382 ms | 1.0x |
| 65×65 | 9.606 ms | 9.558 ms | 1.0x |

TF32 costs nothing anywhere except between roughly 25 and 41, where it costs about 11x. And `sigma=4` gives exactly 33. A 65×65 kernel does four times the arithmetic and still finishes three times faster than the default one.

Large depthwise kernels being awkward through cuDNN is a known rough edge (pytorch/pytorch#21462 and others). What's unlucky here is where the default lands.

Either way the separable form doesn't care: 0.264 ms with TF32 on, 0.266 ms with it off.

### The output stays the same

The two forms differ by `3.99e-04` relative in float32, and by `3.07e-08` in float64. That's floating point accumulation order, not a real difference.

I also ran PatchCore on five MVTec categories with the same trained model, swapping only the blur:

| category | image AUROC diff | pixel AUROC diff |
|---|---|---|
| screw | 0.00e+00 | 1.04e-06 |
| transistor | 0.00e+00 | 8.10e-07 |
| capsule | 0.00e+00 | 1.27e-06 |
| grid | 0.00e+00 | 9.78e-07 |
| cable | 0.00e+00 | 7.80e-07 |

Image AUROC came out identical in all five. Pixel AUROC moves in the sixth decimal.

### What I'd change

It's all inside `src/anomalib/models/components/filters/blur.py`.

In `__init__`, after the 2D kernel is built, get the two 1D kernels out of it by summing along each axis. For a separable kernel that sums to 1, this is exact:

```python
single = self.kernel[:1, :1]
self.register_buffer(
    "kernel_y",
    single.sum(dim=-1, keepdim=True).expand(self.channels, -1, -1, -1).contiguous(),
    persistent=False,
)
self.register_buffer(
    "kernel_x",
    single.sum(dim=-2, keepdim=True).expand(self.channels, -1, -1, -1).contiguous(),
    persistent=False,
)
```

Then in `forward`, one line becomes two:

```python
# currently
output = F.conv2d(input_tensor, self.kernel, groups=self.channels, padding=0, stride=1)

# proposed
output = F.conv2d(input_tensor, self.kernel_x, groups=self.channels, padding=0, stride=1)
output = F.conv2d(output, self.kernel_y, groups=self.channels, padding=0, stride=1)
```

I pulled the 1D kernels out of the existing buffer instead of recomputing them from `sigma` on purpose. That way `self.kernel` stays exactly as it is, the `state_dict` keeps the same keys, and current checkpoints load without any migration. The 1D buffers are `persistent=False` so they never end up in a checkpoint either.

28 lines added, 2 changed. No API change and nothing for users to switch on.

I've got it running against `main` with 17 checks passing: equivalence for sigma 1, 2, 4 and 8, for 1, 3 and 8 channels, for different sigmas per axis, for explicit and non-square `kernel_size`, for `padding="valid"`, and for inputs smaller than the kernel. The `state_dict` keeps its keys and loads from an unpatched model, `.to(device)` moves the 1D buffers, float16 works, gradients flow. Small kernels don't regress: `sigma=1` comes out 1.66x faster, `sigma=2` 3.98x.

`GaussianBlur2d` is used in 13 places, so `cfa`, `dinomaly`, `inp_former`, `padim`, `patchcore`, `reverse_distillation`, `super_add`, `supersimplenet`, `uninet` and the `tiled_ensemble` smoothing all go through it. I only timed PatchCore and PaDiM end to end.

### Things you'll probably want to ask

**Why not drop the blur entirely?** That's cheaper still, and honestly the detection barely notices. I measured pixel AUROC at 0.9812 with no blur against 0.9816 today, averaged over four categories. The problem is it changes what everyone sees. Heatmaps go blocky, saved thresholds shift, and published numbers across ten models would need redoing. What I'm proposing changes nothing and still gets almost all the time back. After the fix the map generator is 0.54 ms inside a 41.70 ms forward, so dropping it altogether would save just over one percent.

Same story for the other cheap options I tried on the same score grids: blur the small grid then upsample (0.104 ms, −0.0000 AUROC), three box passes (0.314 ms, −0.0001), average pooling (0.161 ms, −0.0022). They all work and they all change the output, and once the blur is fixed there's very little left for them to win.

**Why not skip the anomaly map when you only need the score?** For classification that would help, but there's no way to do it today. `visualizer=False`, `post_processor=False`, `evaluator=False`, and all three together still produce the map. `Patchcore(task=...)` raises `TypeError` and `Engine.__init__` doesn't take a `task`. A custom post-processor runs after the forward, so the cost is already spent by then. That might deserve its own issue, but it would only help classification users. Most people doing anomaly detection want to see where the defect is, and segmentation is the default task anyway. Those users pay for the blur too, and this fix helps them.

**Why not turn TF32 off?** It does work, 26.2 ms down to 2.4 ms. But it's a global flag and TF32 is what makes the backbone fast. With it off the feature extractor goes from 9.02 ms to 13.91 ms, so you'd be spending 4.9 ms of backbone time to save 24 ms of blur time. Better to just not run the expensive convolution. The separable form is unaffected by TF32 either way.

**Why not call `kornia.gaussian_blur2d`?** It's already separable and it'd be one line, so I measured it. 0.685 ms, against 0.174 ms for applying the precomputed kernel separably. It rebuilds the kernel on every call, which is the thing this class exists to avoid. Its own docstring says it "pre-computes the Gaussian kernel during initialization for efficiency".

**Would a smaller `sigma` help?** `sigma` and `kernel_size` change the output, so they're modelling choices and not performance knobs. `cudnn.benchmark = True` makes no measurable difference (26.264 ms against 26.000 ms).

### What I haven't checked

- ONNX export and `torch.jit.script`. Two convs with 1×k and k×1 kernels should export fine, but I haven't confirmed it.
- `torch.compile`. Inductor wants Triton and I'm on Windows. I doubt it would help, since turning a 2D convolution into two 1D ones is algebra rather than kernel fusion, but that's a guess and not a measurement.
- Whether your existing tests already cover this class.
- Non-NVIDIA hardware. The 16x from doing less arithmetic should hold anywhere, Intel and AMD included, but the TF32 band is specific to NVIDIA Ampere and later, so I'd expect a smaller total elsewhere. With TF32 off I measured 9.1x on the blur, which is probably the floor. I only have one card to test on.
- CPU, where I measured 1.03x. This is a GPU thing.

It doesn't save meaningful memory either. Peak for the blur goes from 4.4 MB to 1.3 MB.

Happy to open the PR if you think it's worth having, and just as happy to leave it with you. Either way, thanks for anomalib. It's saved me a lot of work.

### Dataset

MVTecAD

### Model

PatchCore

### Steps to reproduce the behavior

The blur is reachable directly, so no training run is needed:

```python
import time
import torch
from anomalib.models.components.filters import GaussianBlur2d

def bench(fn, reps=20, warmup=8):
    for _ in range(warmup):
        fn()
    torch.cuda.synchronize()
    start = time.perf_counter()
    for _ in range(reps):
        fn()
    torch.cuda.synchronize()
    return (time.perf_counter() - start) / reps * 1000

# sigma=4 is the default in AnomalyMapGenerator, and gives a 33x33 kernel
blur = GaussianBlur2d(sigma=4.0, channels=1).cuda().eval()
x = torch.randn(1, 1, 320, 320, device="cuda")

with torch.no_grad():
    print(f"kernel {tuple(blur.kernel.shape[-2:])}: {bench(lambda: blur(x)):.2f} ms")

    torch.backends.cudnn.allow_tf32 = False
    print(f"same, TF32 off:   {bench(lambda: blur(x)):.2f} ms")
    torch.backends.cudnn.allow_tf32 = True
```

To see it inside a model, time a PatchCore forward and then time `model.model.anomaly_map_generator` on its own. On my machine the generator is 25.38 ms of a 60.99 ms forward.

### OS information

OS information:
OS: Windows 11 (build 26200)
Python version: 3.10.10
Anomalib version: 2.2.0 for the measurements. I also checked main at db4e8cb (2.6.2.dev0) and GaussianBlur2d is identical there, so this isn't something already fixed in a newer release.
PyTorch version: 2.5.1+cu121
CUDA/cuDNN version: CUDA 12.1, cuDNN 9.1.0
GPU models and configuration: 1x NVIDIA RTX 1000 Ada Generation Laptop GPU, 6 GB, compute capability 8.9
Any other relevant information: timm 1.0.24, kornia 0.8.2, lightning 2.6.0. Compute capability 8.9 means TF32 is on by default, which accounts for about half of the slowdown described above.

### Expected behavior

A Gaussian blur is separable, so I would expect it to be applied as two 1D passes: cost scaling with 2k per pixel instead of k*k, and no large 2D kernel for cuDNN to pick an algorithm for.
Concretely, at the default sigma=4 and 320x320 I would expect the blur to take well under a millisecond rather than 25 ms, and the anomaly map generator to be a rounding error in the forward pass instead of 40% of it.
The output should be identical either way, since the two forms are the same operation.

### Pip/GitHub

pip

### What version/branch did you use?

2.2.0 from pip for all measurements. I also read main at db4e8cb (2.6.2.dev0) to confirm GaussianBlur2d is unchanged there.

### Configuration YAML

```yaml
No config file involved. The snippet above constructs GaussianBlur2d directly, and the model timings used the Python API with default arguments:
Patchcore(backbone="wide_resnet50_2", layers=["layer2", "layer3"], num_neighbors=9)
Padim(backbone="resnet18", layers=["layer1", "layer2", "layer3"])
```

### Logs

```shell
Output of the snippet above:


kernel (33, 33): 25.92 ms
same, TF32 off:   2.04 ms


Run-to-run variance is around 5%, so the figures move a little, but the ratio holds.

Kernel size sweep on the same machine, plain depthwise conv2d at 320x320:


kernel    TF32 on     TF32 off    ratio
 9x9      0.213 ms    0.213 ms     1.0x
17x17     0.639 ms    0.651 ms     1.0x
25x25    15.486 ms    1.344 ms    11.5x
33x33    26.103 ms    2.443 ms    10.7x
41x41    40.483 ms    3.429 ms    11.8x
49x49     5.513 ms    5.382 ms     1.0x
65x65     9.606 ms    9.558 ms     1.0x
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by andreifagds** [comment]
Re-reading my issue, I softened the ending by saying I'd be just as happy to leave it with you. Honestly, that wasn't entirely true. I'd really like to be the one to do it.

anomalib runs on a PCB solder inspection setup I built on a production line, and it's something I've put a lot of time into. The patch and the tests are ready against main. If this is a solution that makes sense to you, I'd be very happy to open the PR and make my first contribution to anomalib.

**Comment by ashwinvaidya17** [comment]
@andreifagds appreciate the detailed investigation and the explanation. Do you also plan to change the default sigma as it affects the kernel size? From what I understand only 1d conv has the large kernel size issue.

Also, can you add the anomaly maps with these experiments. Like you said, I was concerned about blobs as well when removing gaussian2d. I am curious on the effect on visual quality of the generated maps. It is also possible that we might not have the best defaults but where possible we've tried to match the params to the paper/official code.

In either case feel free to create a PR and we can take it from there.

And just to explain the rationale we moved away from Kornia as we were facing issues exporting to onnx/openvino due to kornia kernels. It might have changed meanwhile. I haven't checked in a while.

**Comment by andreifagds** [comment]
Thanks for looking at it so quickly, and for the green light.

Starting with the maps, since that was the main question. Three MVTec
categories, the worst-scoring defective image in each, same trained model, only
the blur swapped:

<img width="1693" height="1151" alt="Image" src="https://github.com/user-attachments/assets/f74d45eb-0489-4fb2-818d-cdabd9eb257b" />


The worst difference is 7.6e-03 on maps that go up to around 80, so it looks
like float32 accumulation order and nothing else. Columns 2 and 3 are the same
picture.

**On the sigma** — I'd rather leave it alone, if that's alright with you. The
kernel stays 33×33 with every weight exactly where it is, and only the way it
gets applied changes. You mentioned the defaults come from the paper, and that
seemed like a good reason not to touch them.

**On which conv is slow** — I explained this badly in the issue, sorry. It's the
2D one. The 1D ones seem not to care about TF32 at all:

| | TF32 on | TF32 off |
|---|---|---|
| one conv2d, 33×33 kernel | 26.10 ms | 2.44 ms |
| two conv2d, 33×1 and 1×33 | 0.264 ms | 0.266 ms |

The bad band looks like it sits between roughly 25 and 41 elements per side, and
a 33×1 kernel never reaches it.

**On the blobs** — you were right to be suspicious of what I wrote there. I said
removing the blur would make the maps blocky, and that was an overstatement. The
bilinear upsample already smooths a fair bit, so column 5 comes out grainy
rather than blocky, and pixel AUROC only moves from 0.9816 to 0.9812 across four
categories.

So yes, you could just remove it, and the speed would land in about the same
place either way, since both routes drop the same 25 ms. What made me hesitate
is that the maps would then change for everyone already using it, their saved
thresholds would shift, and it would move away from the paper. I'd guess that
isn't what you'd want, though you'd know that better than me. Does the version
where nothing changes seem like the right trade, or would you rather have the
conversation about dropping the blur properly? Either way there's under 1% left
in it after this, since the map generator comes down to 0.54 ms inside a
41.70 ms forward.

**On kornia and export** — that's really useful to know, and it makes me glad I
didn't go the `kornia.gaussian_blur2d` route. The change stays inside
`GaussianBlur2d`, the forward only uses `F.conv2d`, and the 1D kernels come from
summing the existing 2D buffer along each axis, so there's nothing new from
kornia at runtime. I tested both export paths:

| | current | separable |
|---|---|---|
| ONNX opset 17, `onnx.checker.check_model` | valid | valid |
| PyTorch vs ONNX | 0.000e+00 | 0.000e+00 |
| OpenVINO `convert_model` + CPU plugin | works | works |
| PyTorch vs OpenVINO | 0.000e+00 | 0.000e+00 |
| operators | Conv, Pad, Resize, Slice, Concat, Transpose, Reshape, Cast, Shape, Constant, ConstantOfShape | same set |
| onnxruntime CPU, map generator | 6.94 ms | 1.76 ms |
| OpenVINO CPU, map generator | 1.44 ms | 0.34 ms |

Same operator set either way, and both runtimes come out faster too. Those are
CPU numbers with no TF32 involved, so that part is purely the reduced
arithmetic.

I'll go ahead and open the PR then. Thanks again for taking the time on this,
it's my first contribution here and it means a lot that you read it properly.

**Comment by andreifagds** [comment]
An update on this, and an apology: my numbers don't hold.

I rebuilt everything in a clean environment before opening the PR, and found
that my measurements came from torch 2.5.1, which is below the `torch>=2.6.0`
minimum. The 25 ms is a cuDNN behaviour and it moves a lot between versions. On
the 9.10.2 that `uv.lock` pins, the same convolution is already down to 0.38 ms,
which is why the number looked implausible to you.

Making the blur separable does still help the operation itself, around 2.5x to
9.6x depending on shape, and it comes out slightly closer to the exact result
since the thin kernels don't fall into TF32. But end to end it's within noise at
256×256 with a single image, and only reaches 10 to 15% with a lighter backbone
or a larger input. Against a size threshold, a second code path and an exported
graph that roughly doubles in nodes, I don't think that's a trade worth asking
you to maintain, so I'd rather withdraw it than push it through.

Sorry for the noise. Feel free to close this one.

Thanks for engaging with it as carefully as you did, by the way. Your question
about the anomaly maps is what sent me to rebuild the environment properly, and
without it I'd have opened a PR with numbers that don't reproduce.

I'd like to keep looking for something more worthwhile here. If there's an area
where profiling work would actually be welcome, I'd be glad to hear it.

**Comment by ashwinvaidya17** [comment]
All good. Thanks for looking into a potential issue. I'll close the issue.
